Adding migration of GitHub Secret Scanning Alerts

peter-murray opened the pull request:
<!--
For the checkboxes below you must check each one to indicate that you either did the relevant task, or considered it and decided there was nothing that needed doing
-->

- [x] Did you write/update appropriate tests
- [x] Release notes updated (if appropriate)
- [x] Appropriate logging output
- [x] Issue linked
- [x] Docs updated (or issue created) (#715 )

This is the first pass on the requirements to migrate secret scanning remediations between two repositories, provides the functionality for #536.

This is being opened to start a code review and then complete the support of this command line and service. There are some necessary manual steps required as part of the larger migration in that the repo is migrated and secret scanning has been reactivated (and the backfill scan completed) before this command line makes sense to execute.

Created a new issue to add an integration test to cover this functionality in the future: #717 

<!--
For docs we should review the docs at:
https://docs.github.com/en/early-access/github/migrating-with-github-enterprise-importer
and the README.md in this repo

If a doc update is required based on the changes in this PR, it is sufficient to create an issue and link to it here. The doc update can be made later/separately.

The process to update the docs can be found here:
https://github.com/github/docs-early-access#opening-prs

The markdown files are here: 
https://github.com/github/docs-early-access/tree/main/content/github/migrating-with-github-enterprise-importer
-->

ArinGhazarian commented on src/Octoshift/GithubClient.cs:
```diff
@@ -111,27 +114,27 @@ public virtual async IAsyncEnumerable<JToken> GetAllAsync(string url)
             _log.LogVerbose($"HTTP {httpMethod}: {url}");
 
-            if (body != null)
+            using var request = new HttpRequestMessage(httpMethod, new Uri(url));
```
I would follow `HttpClient`'s implementation and use the following method to create the `Uri`:
```c#
private Uri CreateUri(string? uri) => string.IsNullOrEmpty(uri) ? null : new Uri(uri, UriKind.RelativeOrAbsolute);
```

ArinGhazarian replied:
If you get the latest main, this method is already rewritten.

ArinGhazarian commented on src/Octoshift/GithubClient.cs:
```diff
@@ -111,27 +114,27 @@ public virtual async IAsyncEnumerable<JToken> GetAllAsync(string url)
+                _log.LogVerbose($"HTTP BODY: {body.ToJson()}");
+                
+                var payload = body.ToJson();
+                request.Content = new StringContent(payload, Encoding.UTF8, "application/json");
```
We already have an extension for this. Also since we are instantiating the `HttpContent` we need to dispose it:
```c#
using var payload = body?.ToJson().ToStringContent();
request.Content = payload;
```

peter-murray replied:
Don't we lose the encoding and the media type of the content when doing this?

peter-murray replied:
I can see that the headers in the request automatically set to JSON after making the change.

ArinGhazarian commented on src/Octoshift/SecretScanningAlertService.cs:
```diff
@@ -0,0 +1,163 @@
+using Octoshift.Models;
+using OctoshiftCLI;
+
+namespace Octoshift
```
nit: use file scoped namespace to reduce indentation.

ArinGhazarian commented on src/gei/Commands/MigrateSecretScanningAlertsCommand.cs:
```diff
@@ -0,0 +1,195 @@
+            var sourceGitHubApi = _sourceGithubApiFactory.Create(args.GhesApiUrl, args.GithubSourcePat);
+            var targetGithubApi = _targetGithubApiFactory.Create(args.TargetApiUrl, args.GithubTargetPat);
+
+            var migrationService = new SecretScanningAlertService(sourceGitHubApi, targetGithubApi, _log);
```
We use DI so the `SecretScanningAlertService` defiantly needs to be injected so that it can be mocked in tests. Now since it requires the source and target APIs, you are going to need to create a Factory and inject the factory into the command (eg. `secretScanningAlertServiceFactory.Create(sourceGitHubApi, targetGithubApi, _log)`)

ArinGhazarian commented on src/gei/Commands/MigrateSecretScanningAlertsCommand.cs:
```diff
@@ -0,0 +1,195 @@
+using Octoshift;
+using OctoshiftCLI.Extensions;
+
+namespace OctoshiftCLI.GithubEnterpriseImporter.Commands
```
nit: use file scoped namespace.

ArinGhazarian commented on src/Octoshift/Models/CodeScanningAnalysis.cs:
```diff
@@ -0,0 +1,34 @@
+namespace Octoshift.Models
+{
+    public class CodeScanningAnalysis
```
So far in most of our code we were able to escape creating DTOs just by using dynamic json objects. So I guess it would be possible to do so here too and it probably applies to all models that you created.

ArinGhazarian commented on src/Octoshift/SecretScanningAlertService.cs:
```diff
@@ -0,0 +1,163 @@
+            {
+                _log.LogInformation($"Processing source secret {alert.Alert.Number}");
+                
+                if (alert.Alert.State == "resolved")
```
Here also I would create `const`s instead of using magic strings. For these constants you might also want to follow what we did for [RepositoryMigrationStatus](https://github.com/github/gh-gei/blob/8736d28aef48fb4cd7d784e1f7328d1b27470883/src/Octoshift/RepositoryMigrationStatus.cs#L3) if you intend to use them in other places.

ArinGhazarian commented on src/Octoshift/SecretScanningAlertService.cs:
```diff
@@ -0,0 +1,163 @@
+                        }
+                    }
+                }
+                else
+                {
+                    _log.LogInformation("  secret alert is still open, nothing to do");    
+                }
```
IMO it's better to invert the if-else here in favor of less indentation:
```c#
if (state is not "resolved")
{
    _log.LogInformation(" secret alert is still open, nothing to do");
    return;
}

... 
```

peter-murray replied:
This is inside a for loop, return would stop the processing of the subsequent secret alerts.

ArinGhazarian replied:
Oh yeah sorry my bad. So then the `return` should be `continue`.

ArinGhazarian commented on src/Octoshift/SecretScanningAlertService.cs:
```diff
@@ -0,0 +1,163 @@
+                {
+                    _log.LogInformation("  secret alert is still open, nothing to do");    
+                }
+                _log.LogInformation($"");
```
Is this intentional?

peter-murray replied:
Yes, it separated the log outputs on the processing of each secret, but I have indented all the other log lines to achieve the desired outcome.

ArinGhazarian commented on src/gei/Commands/MigrateSecretScanningAlertsCommand.cs:
```diff
@@ -0,0 +1,195 @@
+
+namespace OctoshiftCLI.GithubEnterpriseImporter.Commands
+{
+    public class MigrateSecretScanningAlertsCommand : Command
```
This new command class also needs unit tests.

ArinGhazarian reviewed (changes requested):
Overall looks very good, I just added a few comments and suggested some minor changes.

peter-murray commented:
I rebased the changes on the the current head and addressed the various comments from the review in this commit.

I have left the DTO objects in place for now (https://github.com/github/gh-gei/pull/592#discussion_r953038706) as these are useful for building the processing logic and this is the first of the two/maybe three command lines that may need these objects. The are not completely flat objects either and have optional complex nested parts that are relevant if we need to go deeper into performing matches above the simpler logic that is in place now.

ArinGhazarian commented on src/Octoshift/GithubApi.cs:
```diff
@@ -685,6 +686,95 @@ public virtual async Task<MannequinReclaimResult> ReclaimMannequin(string orgId,
+            
+            var url = $"{_apiUrl}/repos/{org}/{repo}/secret-scanning/alerts/{alertNumber}";
+
+            dynamic payload;
```
`payload` should be an `object`:
```suggestion
            object payload;
```

ArinGhazarian commented on src/Octoshift/CodeScanningAlerts.cs:
```diff
@@ -0,0 +1,20 @@
+namespace OctoshiftCLI
+{
+    public static class CodeScanningAlerts
```
It's more customary for enums (and enum likes) to be singular as opposed to plural:
```suggestion
    public static class CodeScanningAlert
```

peter-murray replied:
I ended up with naming collisions with this and the DTO objects, this also applied to the SecretScanningAlert class that was the same, hence why I was using the pural here. I ended up renaming the DTO objects to be prefixed with Github as in the future there could be other objects from other Git hosting systems that need models.

ArinGhazarian commented on src/Octoshift/SecretScanningAlertService.cs:
```diff
@@ -0,0 +1,169 @@
+            }
+            else
+            {
+                _log.LogInformation("  secret is resolved, looking for matching detection in target...");
+                var target = MatchTargetSecret(alert, targetAlerts);
+
+                if (target == null)
+                {
+                    _log.LogWarning(
+                        $"  failed to locate a matching secret to source secret {alert.Alert.Number} in {targetOrg}/{targetRepo}");
+                }
+                else
+                {
+                    _log.LogInformation(
+                        $"  source secret alert matched alert to {target.Alert.Number} in {targetOrg}/{targetRepo}.");
+
+                    if (alert.Alert.Resolution == target.Alert.Resolution
+                        && alert.Alert.State == target.Alert.State)
+                    {
+                        _log.LogSuccess("  source and target alerts are already aligned.");
+                    }
+                    else
+                    {
+                        _log.LogInformation(
+                            $"  updating target alert:{target.Alert.Number} to state:{alert.Alert.State} and resolution:{alert.Alert.Resolution}");
+
+                        if (dryRun)
+                        {
+                            _log.LogInformation(
+                                $"  executing in dry run mode! Secret alert, {target.Alert.Number}, in repository {targetOrg}/{targetRepo} would have been updated to resolution, {alert.Alert.Resolution}");
+                        }
+                        else
+                        {
+                            await _targetGithubApi.UpdateSecretScanningAlert(targetOrg, targetRepo, target.Alert.Number,
+                                alert.Alert.State, alert.Alert.Resolution);
+                            _log.LogSuccess(
+                                $"  source and target alert state and resolution have been aligned to {alert.Alert.Resolution}.");
+                        }
+                    }
+                }   
```
there is too much `if-else` here and it makes it hard to follow the code. The indentations can be greatly reduced by checking for a condition and `continue`, for example:
```c#
if (SecretScanningAlerts.IsOpen(alert.Alert.State))
{
    _log.LogSuccess(" secret alert is still open, nothing to do");
    continue:
}

_log.LogInformation(" secret is resolved, looking for matching detection in target...");

...

if (target is null)
{
    _log.LogWarning(...);
    continue;
}

...
```

peter-murray replied:
Acknowledged and will refactor

ArinGhazarian commented on src/Octoshift/SecretScanningAlertService.cs:
```diff
@@ -0,0 +1,169 @@
+
+        // We cannot guarantee the ordering of things with the locations and the APIs, typically they would match, but cannot be sure
+        // so we need to iterate over all the targets to ensure a match
+        foreach (var targetLocation in targetLocations)
+        {
+            var targetDetails = targetLocation.Details;
+
+            if (sourceDetails.Path == targetDetails.Path
+                && sourceDetails.StartLine == targetDetails.StartLine
+                && sourceDetails.EndLine == targetDetails.EndLine
+                && sourceDetails.StartColumn == targetDetails.StartColumn
+                && sourceDetails.EndColumn == targetDetails.EndColumn
+                && sourceDetails.BlobSha == targetDetails.BlobSha)
+                // Technically this wil hold, but only if there is not commmit rewriting going on, so we need to make this last one optional for now
+                // && sourceDetails.CommitSha == targetDetails.CommitSha)
+            {
+                return true;
+            }
+        }
+
+        return false;
```
this entire `foreach` can be replaced by a LINQ method:
```c#
var sourceDetails = sourceLocation.Details;
return targetLocations.Any(loc => sourceDetails.Path == loc.Details.Path && ...);
```

ArinGhazarian commented on src/Octoshift/SecretScanningAlerts.cs:
```diff
@@ -0,0 +1,23 @@
+namespace OctoshiftCLI
+{
+    public static class SecretScanningAlerts
```
Same concern, it's more customary to use singular class name here:
```suggestion
    public static class SecretScanningAlert
```

ArinGhazarian commented on src/Octoshift/SecretScanningAlerts.cs:
```diff
@@ -0,0 +1,23 @@
+        public static bool IsOpen(string alertState) => alertState?.Trim().ToLower() is AlertStateOpen;
+
+        public static bool IsOpenOrResolved(string alertState) =>
+            alertState?.Trim().ToLower() is (AlertStateOpen or AlertStateResolved);
```
nit: remove the parens

ArinGhazarian commented on src/Octoshift/SecretScanningAlerts.cs:
```diff
@@ -0,0 +1,23 @@
+            alertState?.Trim().ToLower() is (AlertStateOpen or AlertStateResolved);
+        
+        public static bool IsValidDismissedReason(string reason) =>
+            reason?.Trim().ToLower() is (ResolutionFalsePositive or ResolutionRevoked or ResolutionWontFix or ResolutionUsedInTests);
```
nit: remove the parens

ArinGhazarian commented on src/gei/Commands/MigrateSecretScanningAlertsCommand.cs:
```diff
@@ -0,0 +1,196 @@
+
+        LogAndValidateOptions(args);
+
+        var sourceGitHubApi = _sourceGithubApiFactory.Create(args.GhesApiUrl, GetSourceToken(args));
+        var targetGithubApi = _targetGithubApiFactory.Create(args.TargetApiUrl, GetTargetToken(args));
+        var migrationService = _secretScanningAlertServiceFactory.Create(sourceGitHubApi, targetGithubApi);
```
Three is no need to inject source and target Github APIs just to create the `secretScanningAlertService`, the factory can easily create everything that is needed so you can use it like this:
```c#
var migrationService = _secretScanningAlertServiceFactory.Create(args.GhesApiUrl, GetSourceToken(args), args.TargetApiUrl, GetTargetToken(args));
```

ArinGhazarian commented on src/gei/GithubSecretScanningAlertServiceFactory.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    private readonly OctoLogger _octoLogger;
+
+    public GitHubSecretScanningAlertServiceFactory(OctoLogger octoLogger)
+    {
+        _octoLogger = octoLogger;
+    }
```
you can inject the Github API factories and then use them in the `Create` method:

```c#
public GitHubSecretScanningAlertServiceFactory(OctoLogger octoLogger, ISourceGithubApiFactory sourceGithubApiFactory, ITargetGithubApiFactory targetGithubApiFactory)
{
    _octoLogger = octoLogger;
    _sourceGithubApiFactory = sourceGithubApiFactory;
    _targetGithubApiFactory = targetGithubApiFactory;
}
```

ArinGhazarian commented on src/gei/GithubSecretScanningAlertServiceFactory.cs:
```diff
@@ -0,0 +1,15 @@
+        _octoLogger = octoLogger;
+    }
+    
+    public SecretScanningAlertService Create(GithubApi sourceApi, GithubApi targetApi) => new (sourceApi, targetApi, _octoLogger);
```
```c#
SecretScanningAlertService Create(string sourceApiUrl, string sourceApiUrl, string targetApiUrl, string targetPersonalAccessToken)
{
    var sourceGithubApi = _sourceGithubApiFactory.Create(sourceApiUrl, sourceApiUrl);
    var targetGithubApi = _targetGithubApiFactory.Create(targetApiUrl, targetPersonalAccessToken);
    ...
```
Note: You might also want NoSSL

ArinGhazarian commented on src/gei/IGithubSecretScanningAlertServiceFactory.cs:
```diff
@@ -0,0 +1,9 @@
+using Octoshift;
+
+namespace OctoshiftCLI.GithubEnterpriseImporter
```
nit: use file scoped namespace.

ArinGhazarian commented on src/gei/Program.cs:
```diff
@@ -36,6 +37,7 @@ public static async Task Main(string[] args)
                 .AddSingleton<IVersionProvider, VersionChecker>(sp => sp.GetRequiredService<VersionChecker>())
                 .AddTransient<ITargetGithubApiFactory>(sp => sp.GetRequiredService<GithubApiFactory>())
                 .AddTransient<ISourceGithubApiFactory>(sp => sp.GetRequiredService<GithubApiFactory>())
+                .AddTransient<ISecretScanningAlertServiceFactory>(sp => sp.GetRequiredService<GitHubSecretScanningAlertServiceFactory>())
```
There is no need to resolve the `GitHubSecretScanningAlertServiceFactory` here, so the registration can be simplified:
```suggestion
                .AddTransient<ISecretScanningAlertServiceFactory, GitHubSecretScanningAlertServiceFactory>()
```

The reason that `ITargetGithubApiFactory` is registered like that is to resolved the same `GithubApiFactory`. (we are going to do a refactor later on to register everything as a singleton, which will simplify the registrations)

ArinGhazarian commented on src/gei/Program.cs:
```diff
@@ -27,6 +27,7 @@ public static async Task Main(string[] args)
                 .AddSingleton(Logger)
                 .AddSingleton<EnvironmentVariableProvider>()
                 .AddSingleton<GithubApiFactory>()
+                .AddSingleton<GitHubSecretScanningAlertServiceFactory>()
```
In favor of a simpler registration, this line can be removed:
```suggestion
```

ArinGhazarian commented on src/OctoshiftCLI.Tests/gei/Commands/MigrateSecretScanningAlertsCommandTests.cs:
```diff
@@ -0,0 +1,59 @@
+using OctoshiftCLI.GithubEnterpriseImporter.Commands;
+using Xunit;
+
+namespace OctoshiftCLI.Tests.GithubEnterpriseImporter.Commands
```
nit: use file scoped namespace.

ArinGhazarian commented on src/OctoshiftCLI.Tests/gei/Commands/MigrateSecretScanningAlertsCommandTests.cs:
```diff
@@ -0,0 +1,59 @@
+            TestHelpers.VerifyCommandOption(_command.Options, "verbose", false);
+            TestHelpers.VerifyCommandOption(_command.Options, "dry-run", false);
+        }
+    }
```
I'd also like to see a test covering a happy path (or a sad path if any) scenario.

peter-murray replied:
Ouch, the amount of data required to mock that via the GitHub APIs is pretty heavy to get it all to link up... I will do so but maintaining that test is bound to be pretty horrible if it ever needs changing.

ArinGhazarian replied:
Usually it's a code smell whenever we need to mock too much. Let's start with having a test and later on we can refactor in a separate PR if need be.

peter-murray replied:
Understood, but its the underlying data from the source and target here needs to be carefully mocked to cover the matching between the secret alerts and the various outcomes that can arise. I can do a trivial one, but might as well do something meaningful if going down this path.

ArinGhazarian replied:
> but its the underlying data from the source and target here needs to be carefully mocked to cover the matching between the secret alerts and the various outcomes that can arise

Yes. This is exactly the reason why we add different unit tests and spend some time trying to cover edge cases, etc., which will save us a lot of time in the future in case of any bugs, etc. So I would spend a bit more time writing good tests to test all scenarios to prevent future bugs that could have been caught by tests.

peter-murray commented on src/Octoshift/GithubApi.cs:
```diff
@@ -3,6 +3,7 @@
 using System.Linq;
 using System.Net;
 using System.Net.Http;
+using System.Net.Http.Headers;
```
```suggestion
```

peter-murray commented on src/Octoshift/GithubApi.cs:
```diff
@@ -685,6 +686,95 @@ public virtual async Task<MannequinReclaimResult> ReclaimMannequin(string orgId,
+            var url = $"{_apiUrl}/repos/{org}/{repo}/code-scanning/alerts/{alertNumber}";
+
+            object payload;
+            if (state == "open")
```
Sure thing, I missed that one, there is a constant for this.

ArinGhazarian commented on src/Octoshift/GithubApi.cs:
```diff
@@ -685,6 +685,81 @@ public virtual async Task<MannequinReclaimResult> ReclaimMannequin(string orgId,
+
+            var url = $"{_apiUrl}/repos/{org}/{repo}/code-scanning/alerts/{alertNumber}";
+
+            var payload = state == CodeScanningAlert.AlertStateOpen
+                ? (new { state })
+                : (object)(new
+                {
+                    state,
+                    dismissed_reason = dismissedReason,
+                    dismissed_comment = dismissedComment
+                });
```
nit:
```suggestion
            object payload = state == CodeScanningAlert.AlertStateOpen
                ? new { state }
                : new
                {
                    state,
                    dismissed_reason = dismissedReason,
                    dismissed_comment = dismissedComment
                };
```

ArinGhazarian commented on src/Octoshift/GithubApi.cs:
```diff
@@ -685,6 +685,81 @@ public virtual async Task<MannequinReclaimResult> ReclaimMannequin(string orgId,
+
+            var url = $"{_apiUrl}/repos/{org}/{repo}/secret-scanning/alerts/{alertNumber}";
+
+            var payload = state == SecretScanningAlert.AlertStateOpen ? (new { state }) : (object)(new { state, resolution });
```
nit:
```suggestion
            object payload = state == SecretScanningAlert.AlertStateOpen ? new { state } : new { state, resolution };
```

ArinGhazarian commented on src/Octoshift/Models/GithubCodeScanningAnalysis.cs:
```diff
@@ -0,0 +1,34 @@
+namespace Octoshift.Models
```
nit: use file scoped namespace

ArinGhazarian commented on src/Octoshift/Models/GithubSecretScanningAlert.cs:
```diff
@@ -0,0 +1,38 @@
+namespace Octoshift.Models
```
nit: use file scoped namespace

ArinGhazarian commented on src/Octoshift/SecretScanningAlert.cs:
```diff
@@ -0,0 +1,23 @@
+namespace OctoshiftCLI
```
nit: use file scoped namespace

ArinGhazarian commented on src/Octoshift/SecretScanningAlertService.cs:
```diff
@@ -0,0 +1,165 @@
+    private bool IsMatchedSecretAlertLocation(GithubSecretScanningAlertLocation sourceLocation,
+        GithubSecretScanningAlertLocation[] targetLocations)
+    {
+        var sourceDetails = sourceLocation.Details;
+
+        // We cannot guarantee the ordering of things with the locations and the APIs, typically they would match, but cannot be sure
+        // so we need to iterate over all the targets to ensure a match
+        foreach (var targetLocation in targetLocations)
+        {
+            var targetDetails = targetLocation.Details;
+
+            if (sourceDetails.Path == targetDetails.Path
+                && sourceDetails.StartLine == targetDetails.StartLine
+                && sourceDetails.EndLine == targetDetails.EndLine
+                && sourceDetails.StartColumn == targetDetails.StartColumn
+                && sourceDetails.EndColumn == targetDetails.EndColumn
+                && sourceDetails.BlobSha == targetDetails.BlobSha)
+            // Technically this wil hold, but only if there is not commmit rewriting going on, so we need to make this last one optional for now
+            // && sourceDetails.CommitSha == targetDetails.CommitSha)
+            {
+                return true;
+            }
+        }
+
+        return false;
```
`foreach` can be replaced by a simple LINQ method, see my [previous comment](https://github.com/github/gh-gei/pull/592#discussion_r964366163)

ArinGhazarian commented on src/gei/Commands/MigrateSecretScanningAlertsCommand.cs:
```diff
@@ -0,0 +1,196 @@
+
+    public MigrateSecretScanningAlertsCommand(OctoLogger log,
+        ISecretScanningAlertServiceFactory secretScanningAlertServiceFactory,
+        ISourceGithubApiFactory sourceGithubApiFactory,
+        ITargetGithubApiFactory targetGithubApiFactory,
```
No need to inject these factories anymore.

ArinGhazarian commented on src/gei/GithubSecretScanningAlertServiceFactory.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace OctoshiftCLI.GithubEnterpriseImporter;
+
+public sealed class GitHubSecretScanningAlertServiceFactory : ISecretScanningAlertServiceFactory
```
We made a decision not to create namespaces unless it's absolutely necessary. We create ordinary classes and mark the public methods as `virtual` to be able to mock them.  
```suggestion
public class GitHubSecretScanningAlertServiceFactory
```

ArinGhazarian commented on src/gei/GithubSecretScanningAlertServiceFactory.cs:
```diff
@@ -0,0 +1,29 @@
+        _targetGithubApiFactory = targetGithubApiFactory;
+    }
+
+    public SecretScanningAlertService
```
```suggestion
    public virtual SecretScanningAlertService
```

ArinGhazarian commented on src/gei/IGithubSecretScanningAlertServiceFactory.cs:
```diff
@@ -0,0 +1,7 @@
+using Octoshift;
+
+namespace OctoshiftCLI.GithubEnterpriseImporter;
+public interface ISecretScanningAlertServiceFactory
```
Based on what I have already explained, this interface can be deleted.

ArinGhazarian reviewed (changes requested):
Most of the comments were addressed and it's looking already better. There are still some things that need to be addressed including fixing the linting problems by running `dotnet format`

peter-murray commented:
@ArinGhazarian when I run `dotnet format` I am only getting a single response about CA1720, is there something I am missing from this, that was invoking it from the `src` directory?
```
$ dotnet format
Unable to fix CA1720. No associated code fix found.
```

ArinGhazarian commented:
> @ArinGhazarian when I run `dotnet format` I am only getting a single response about CA1720, is there something I am missing from this, that was invoking it from the `src` directory?
> 
> ```
> $ dotnet format
> Unable to fix CA1720. No associated code fix found.
> ```

It is a compiler warning () and it's because there is a property named as `Guid` (which is guess is a DTO):
`Warning: /home/runner/work/gh-gei/gh-gei/src/Octoshift/Models/GithubCodeScanningAnalysis.cs(29,19): warning CA1720: Identifier 'Guid' contains type name`. I guess you can suppress the warning for that one property:
```c#
#pragma warning disable CA1720
// The code that's violating the rule is on this line.
#pragma warning restore CA1720
```

P.S: I still believe that creating those strongly typed DTOs is a bit of an overkill since most of those properties won't be used anyways.

ArinGhazarian reviewed (approved):
Looks good 👍  I would also ask @dylan-smith's review before merging and of course CI needs to green 🟢

peter-murray commented:
The code scanning cli tool that David is working on will be using the model objects to link up and detect the right scanning results to export and import, so I have suppressed the checks on that model object as indicated.

peter-murray commented:
As for the CI build failures, has there been some changes in the main branch that I rebased on in this area as the failures seem unrelated to the code changes I have directly made here?

ArinGhazarian commented on src/Octoshift/Models/GithubCodeScanningAnalysis.cs:
```diff
@@ -0,0 +1,33 @@
+    public string Name { get; set; }
+    public string Guid { get; set; }
+    public string Version { get; set; }
+#pragma warning enable CA1720
```
it should be `restore` not `enable` that's why you're getting the new compiler warning
```suggestion
#pragma warning restore CA1720
```

ArinGhazarian commented:
> As for the CI build failures, has there been some changes in the main branch that I rebased on in this area as the failures seem unrelated to the code changes I have directly made here?

INT tests are a bit flaky today, I would re-run the jobs.

peter-murray commented on src/Octoshift/Models/GithubCodeScanningAnalysis.cs:
```diff
@@ -25,7 +25,9 @@ public class GithubCodeScanningEnvironment
     public string Name { get; set; }
     public string Guid { get; set; }
     public string Version { get; set; }
+#pragma warning enable CA1720
```
It was also complaining about version and name locally when I ran it

peter-murray commented:
So the tests are now passing, but when I run the command locally now, from Rider, the `--github-source-org` even when specified on the command line, is not being passed through to the service. It is marked as required in the options and the options parsing is not failing, but the value is empty and I am capturing it as empty here; https://github.com/github/gh-gei/pull/592/files#diff-ae9bc09e955cefc7c22a65ae1d74f07c2281734b8efc099e9ee9fe53d53d412cR118 and failing

Looking at this and comparing to other commands I cannot see what is different, and the other parameters are coming through fine, it is just the source and target organisation names are not coming through from the command line?

ArinGhazarian commented:
> So the tests are now passing, but when I run the command locally now, from Rider, the `--github-source-org` even when specified on the command line, is not being passed through to the service. It is marked as required in the options and the options parsing is not failing, but the value is empty and I am capturing it as empty here; https://github.com/github/gh-gei/pull/592/files#diff-ae9bc09e955cefc7c22a65ae1d74f07c2281734b8efc099e9ee9fe53d53d412cR118 and failing
> 
> Looking at this and comparing to other commands I cannot see what is different, and the other parameters are coming through fine, it is just the source and target organisation names are not coming through from the command line?
> 
> If I rename the parameter from `--github-source-org` (which is what is used elsewhere) to `--source-org` it works, so have we got something special in place for `--github` prefixed options now?

The way model binding works in the version of `System.CommandLine` we're using is by convention which matches the `Option` name with the `Arg` name, meaning that they need to match. So you need to either rename the `SourceOrg` in the `MigrateSecretScanningAlertsCommandArgs` model to `GithubSourceOrg` or as you already mentioned do it the other way around. Now given the fact that in all of our other command we used `--source-org` we should do the same here and just rename the `Option` name to be `--srouce-org`.

peter-murray commented:
I am happy to go with source-org, but that was not the case in the examples that I was pointed to to base off of at the start. Thank you for the clarification though and `source-org` does better align in this case.

ArinGhazarian reviewed (approved):
👍

davelosert commented:
@ArinGhazarian : Kindly asking if there is any timeline on when this might be merged?

I am already done with the code for the next step (code scanning alert migration) as defined in #535, but it is based on the work that @peter-murray is doing in this PR, so I am blocked with putting up the PR until this is merged.

dylan-smith commented:
> @ArinGhazarian : Kindly asking if there is any timeline on when this might be merged?
> 
> I am already done with the code for the next step (code scanning alert migration) as defined in #535, but it is based on the work that @peter-murray is doing in this PR, so I am blocked with putting up the PR until this is merged.

Hi @davelosert , Arin is unavailable for the next week or two (on-call rotation), but I'll take a look at this on Thursday. Sorry it's taking so long.

peter-murray commented:
@dylan-smith friendly bump on the progress of getting this merged?

dylan-smith commented:
> @dylan-smith friendly bump on the progress of getting this merged?

I msg'd you on slack on friday, got a bunch of commits rdy to push. Just need permission to push em on the fork.

peter-murray commented:
@dylan-smith sent an invite for you on the source repo so you can push changes

timrogers commented:
@peter-murray Could you provide a bit more context here on why we skip open alerts? I've seen this come up in a customer support ticket.

dylan-smith commented:
> @peter-murray Could you provide a bit more context here on why we skip open alerts? I've seen this come up in a customer support ticket.

It’s because if the alert is open there is nothing to migrate. When a secret scan is done on the target you’ll end up with the same open alert without needing to migrate anything.

we’re only migrating closed state and reason.

timrogers commented:
> > @peter-murray Could you provide a bit more context here on why we skip open alerts? I've seen this come up in a customer support ticket.
> 
> It’s because if the alert is open there is nothing to migrate. When a secret scan is done on the target you’ll end up with the same open alert without needing to migrate anything.
> 
> we’re only migrating closed state and reason.

Thanks! That makes sense. That's the reason I guessed, but I just wanted to validate that.

We should add some documentation for this command to explain a bit more of how it works. We have the built-in CLI documentation, but it doesn't go into much detail, and I've seen that this can confuse customers.
